### PR TITLE
Run benchmarks on ios

### LINKED ISSUE #2140: Run benchmarks on iOS
Benchmarks should be available also on iOS.

### PR DESCRIPTION
**Pull request by papafe**
Added benchmarks for iOS

Some notes

- I had to use a pre-release build of `BenchmarkDotNet` in order to run the benchmarks on Xamarin ([https://github.com/dotnet/BenchmarkDotNet/issues/1430](https://github.com/dotnet/BenchmarkDotNet/issues/1430)). 
- Nevertheless, this version does not support `BenchmarkSwitcher` (that we use for the benchmarks on .NET 5.0 for Windows, Linux and MacOS) for Xamarin, so I had to use `BenchmarkRunner`. Unfortunately this does not support command line arguments (but there is a [PR](https://github.com/dotnet/BenchmarkDotNet/pull/1292) for that), so I had to do some parsing myself. I tried to keep it consistent with the usual arguments for `BenchmarkDotNet`.
- I removed the `InProcessEmitToolchain` toolchain, as it uses reflection and so it's not available on iOS. From the doc I can see that this is used mostly to speed up the benchmarks, but this doesn't seem a concern for us.
- The project is able to run run the benchmarks both from command line and from the UI. At the moment I am not sure of what is the best way to show the results in the UI, so that part is empty.
- At the moment the workflow for iOS is separated from the other, but it could be merged into one.
- I have also modified the workflow that was already there for Linux, Windows and MacOS, to make the different sections more similar to each other, so that, at some point, we could think about merging them and use parameters, to avoid duplicating commands as we have now.

## Description

Fixes #2140

### CODE REVIEW

File: Tests/Benchmarks/Benchmarks/ViewModel/MainPageViewModel.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        public ICommand RunBenchmarksCommand { get; private set; }
+
+        public MainPageViewModel(string[] args)
+        {
+            this.args = args;
+
+            InitCommands();
+        }
+
+        void InitCommands()
```
Comment (nirinchev) [inline Tests/Benchmarks/Benchmarks/ViewModel/MainPageViewModel.cs:34]: This can be manually inlined in the ctor - not a ton of benefit from having a separate method for it + will allow us to remove the setter from `RunBenchmarksCommand`.

**Review by nirinchev — COMMENTED** [review]
Looks good. Mostly cleanup comments from me. Apart from things commented in the code, we can also:

- Remove most Android images - as far as I remember, Android will fallback to the closest dpi image and we don't really care about the visuals of the test project.
- Same for the iOS images - unless we're forced to have them to install, we should probably get rid of them.

**Comment by papafe** [comment]
Almost finished correcting according to comments.
The only thing left is showing the benchmark in the UI, that let me discover a problem with using `Arguments` attribute with the default toolchain. According to the issue nominated in the output, this should have been resolved, but I have a hunch this wasn't fixed for Xamarin. I'm continuing investigating though

**Comment by papafe** [comment]
So, the situation is even more complicated than I expected :)

Regarding Xamarin.Forms, we have a problem:
- `InProcessEmitToolchain` cannot be used because it uses reflection and it's not supported on iOS
- `InProcessNoEmitToolchain` gives the `Arguments are not supported...` error
It seems that actually we need to give up `Arguments` if we want to support Xamarin.Forms (see bottom of [issue](https://github.com/dotnet/BenchmarkDotNet/issues/687). 

Additionally, and unfortunately, it seems that we have a symmetric issues regarding the benchmarks on Windows, Linux and MacOS (run with .NET 5). If we don't use `InProcessEmitToolchain` the benchmarks don't work, as it complains about not finding the weaver. This is kinda expected, as it's reported in the [documentation](https://benchmarkdotnet.org/articles/guides/troubleshooting.html) that it could not work out of the box for projects with non-trivial build configurations.

Summing up:

- If we want to support testing on Xamarin.Forms, we can't use `ArgumentsAttribute` (hoping that this will be the only issue for in-process benchmarks)
- If we want to keep our headaches to a minimum, we need to use `InProcessEmitToolchain` for benchmarks on .NET 5. This cannot be used on Xamarin though, so the results of the benchmarks will be even less comparable.

**Comment by nirinchev** [comment]
I'm not particularly worried about comparing results between platforms. We'll be mostly using these to catch performance regressions and to validate major refactorings. In these cases we only need to compare against previous runs on the same platform, so we're fine.

To be honest, I don't think I have enough context to fully understand the issue, but will be happy to get in a call and brainstorm it a bit.

**Comment by papafe** [comment]
After discussing with @nirinchev, we agreed that:

- It's not important if some of the platforms use a toolchain and some others a different one, as we will use those to catch performance regressions and evaluate benchmarks over time. So we will continue using the `InProcessEmitToolchain` for the .NET 5.0 tests, and the default toolchain for the others (that is `InProcessNoEmitToolchain` for iOS)
- Because `ArgumentsAttribute` does not work on `InProcessNoEmitToolchain` (and so iOS), we have decided to remove it from the tests, and use `ParametersAttribute` instead.

Regarding the second point, I have refactored the `WriteTests` in order to make theme use parameters. It could have been probably refactored in a different way with more/less abstractions though.

I have also tried to run the benchmarks for iOS on both the simulator and a physical device (both of them on Debug with iPhone X) and the results seem to be comparable:

**iPhone device**

|             Type |           Method | ObjectCount | StringSize | BinarySize | StringLength |         Mean |      Error |     StdDev |       Median |    Gen 0 |   Gen 1 |   Gen 2 | Allocated |
|----------------- |----------------- |------------ |----------- |----------- |------------- |-------------:|-----------:|-----------:|-------------:|---------:|--------:|--------:|----------:|
|       **QueryTests** |            **Count** |          **10** |          **?** |          **?** |            **?** |   **189.655 μs** |  **0.4484 μs** |  **0.3745 μs** |   **189.669 μs** |  **23.9258** |  **0.2441** |  **0.2441** |         **-** |
|       QueryTests |            Count |         100 |          ? |          ? |            ? |   201.375 μs |  0.4397 μs |  0.4113 μs |   201.273 μs |  23.9258 |  0.2441 |  0.2441 |         - |
|       QueryTests |            Count |        1000 |          ? |          ? |            ? |   295.568 μs |  0.5530 μs |  0.5173 μs |   295.583 μs |  23.9258 |  0.4883 |  0.4883 |         - |
|       **QueryTests** |        **Enumerate** |          **10** |          **?** |          **?** |            **?** |   **168.550 μs** |  **0.2379 μs** |  **0.2109 μs** |   **168.573 μs** |  **11.9629** |  **0.4883** |  **0.4883** |         **-** |
|       QueryTests |        Enumerate |         100 |          ? |          ? |            ? |   412.033 μs |  0.4590 μs |  0.4069 μs |   412.071 μs |  27.3438 |  2.4414 |  2.4414 |         - |
|       QueryTests |        Enumerate |        1000 |          ? |          ? |            ? | 3,668.518 μs |  7.1698 μs |  6.7066 μs | 3,668.372 μs | 195.3125 | 19.5313 | 19.5313 |         - |
|      **ObjectTests** | **GetPropertyValue** |           **?** |          **?** |          **?** |            **?** |     **7.449 μs** |  **0.0082 μs** |  **0.0076 μs** |     **7.450 μs** |   **0.2594** |  **0.0305** |  **0.0305** |         **-** |
|      StringTests | GetPropertyValue |           ? |         20 |          ? |            ? |     2.733 μs |  0.0023 μs |  0.0020 μs |     2.733 μs |   0.1259 |       - |       - |         - |
|      StringTests | GetPropertyValue |           ? |        100 |          ? |            ? |     2.408 μs |  0.0256 μs |  0.0240 μs |     2.409 μs |   0.4616 |       - |       - |         - |
|      StringTests | GetPropertyValue |           ? |       1000 |          ? |            ? |    13.909 μs |  0.0123 μs |  0.0096 μs |    13.908 μs |   5.3558 |       - |       - |         - |
|      **StringTests** |       **LookupByPK** |           **?** |         **20** |          **?** |            **?** |    **19.865 μs** |  **0.0260 μs** |  **0.0244 μs** |    **19.870 μs** |   **3.0212** |  **0.1221** |  **0.1221** |         **-** |
|      StringTests |       LookupByPK |           ? |        100 |          ? |            ? |    20.135 μs |  0.0305 μs |  0.0285 μs |    20.140 μs |   3.0212 |  0.1221 |  0.1221 |         - |
|      StringTests |       LookupByPK |           ? |       1000 |          ? |            ? |    56.460 μs |  0.5533 μs |  0.4320 μs |    56.331 μs |   2.9907 |  0.1221 |  0.1221 |         - |
|      **ObjectTests** | **SetPropertyValue** |           **?** |          **?** |          **?** |            **?** |     **2.999 μs** |  **0.0022 μs** |  **0.0021 μs** |     **2.998 μs** |        **-** |       **-** |       **-** |         **-** |
|      StringTests | SetPropertyValue |           ? |         20 |          ? |            ? |     6.226 μs |  0.0102 μs |  0.0085 μs |     6.223 μs |        - |       - |       - |         - |
|      StringTests | SetPropertyValue |           ? |        100 |          ? |            ? |     5.909 μs |  0.0035 μs |  0.0027 μs |     5.909 μs |        - |       - |       - |         - |
|      StringTests | SetPropertyValue |           ? |       1000 |          ? |            ? |    15.891 μs |  0.3111 μs |  0.4045 μs |    15.824 μs |        - |       - |       - |         - |
| **BinaryWriteTests** |            **Write** |          **10** |          **?** |        **128** |            **?** |   **443.647 μs** |  **2.3724 μs** |  **2.1030 μs** |   **444.149 μs** |   **0.9766** |       **-** |       **-** |         **-** |
| BinaryWriteTests |            Write |         100 |          ? |        128 |            ? |   464.851 μs |  1.2485 μs |  1.1067 μs |   464.862 μs |   0.9766 |       - |       - |         - |
| BinaryWriteTests |            Write |          10 |          ? |       1024 |            ? |   446.238 μs |  2.6622 μs |  2.2231 μs |   445.508 μs |   0.9766 |       - |       - |         - |
| BinaryWriteTests |            Write |         100 |          ? |       1024 |            ? |   509.513 μs |  1.1278 μs |  1.0549 μs |   509.694 μs |   0.9766 |       - |       - |         - |
| BinaryWriteTests |            Write |          10 |          ? |    1048576 |            ? |   497.320 μs |  0.9176 μs |  0.8134 μs |   497.441 μs |   0.9766 |       - |       - |         - |
| BinaryWriteTests |            Write |         100 |          ? |    1048576 |            ? |   863.254 μs | 19.2266 μs | 56.6900 μs |   835.050 μs |        - |       - |       - |         - |
|  EmptyWriteTests |            Write |           ? |          ? |          ? |            ? |   428.555 μs |  1.3569 μs |  1.0594 μs |   428.851 μs |   0.4883 |       - |       - |         - |
| StringWriteTests |            Write |          10 |          ? |          ? |           20 |   445.914 μs |  1.1902 μs |  1.0551 μs |   446.345 μs |   0.9766 |       - |       - |         - |
| StringWriteTests |            Write |         100 |          ? |          ? |           20 |   461.859 μs |  1.2687 μs |  1.1868 μs |   461.357 μs |   0.9766 |       - |       - |         - |
| StringWriteTests |            Write |          10 |          ? |          ? |          200 |   445.422 μs |  0.6970 μs |  0.6178 μs |   445.505 μs |   0.9766 |       - |       - |         - |
| StringWriteTests |            Write |         100 |          ? |          ? |          200 |   568.953 μs |  2.1415 μs |  1.8984 μs |   569.043 μs |   0.9766 |       - |       - |         - |

**Simulator**
|             Type |           Method | ObjectCount | StringSize | BinarySize | StringLength |         Mean |      Error |     StdDev |       Median |    Gen 0 |   Gen 1 |   Gen 2 | Allocated |
|----------------- |----------------- |------------ |----------- |----------- |------------- |-------------:|-----------:|-----------:|-------------:|---------:|--------:|--------:|----------:|
|       **QueryTests** |            **Count** |          **10** |          **?** |          **?** |            **?** |   **212.091 μs** |  **4.2158 μs** |  **9.7707 μs** |   **216.083 μs** |  **17.8223** |  **0.2441** |  **0.2441** |         **-** |
|       QueryTests |            Count |         100 |          ? |          ? |            ? |   219.350 μs |  4.3401 μs | 10.1447 μs |   223.307 μs |  17.8223 |  0.2441 |  0.2441 |         - |
|       QueryTests |            Count |        1000 |          ? |          ? |            ? |   260.973 μs |  5.1265 μs |  8.8429 μs |   257.366 μs |  18.0664 |       - |       - |         - |
|       **QueryTests** |        **Enumerate** |          **10** |          **?** |          **?** |            **?** |   **198.419 μs** |  **3.9561 μs** |  **7.9007 μs** |   **201.390 μs** |  **12.4512** |  **0.2441** |  **0.2441** |         **-** |
|       QueryTests |        Enumerate |         100 |          ? |          ? |            ? |   359.997 μs |  7.0516 μs | 10.9785 μs |   358.154 μs |  28.8086 |  1.4648 |  1.4648 |         - |
|       QueryTests |        Enumerate |        1000 |          ? |          ? |            ? | 2,624.653 μs | 51.9092 μs | 57.6969 μs | 2,629.489 μs | 238.2813 | 11.7188 | 11.7188 |         - |
|      **ObjectTests** | **GetPropertyValue** |           **?** |          **?** |          **?** |            **?** |     **6.101 μs** |  **0.1203 μs** |  **0.2138 μs** |     **6.173 μs** |   **0.2670** |  **0.0229** |  **0.0229** |         **-** |
|      StringTests | GetPropertyValue |           ? |         20 |          ? |            ? |     2.949 μs |  0.0121 μs |  0.0094 μs |     2.947 μs |   0.1297 |       - |       - |         - |
|      StringTests | GetPropertyValue |           ? |        100 |          ? |            ? |     5.134 μs |  0.0331 μs |  0.0294 μs |     5.137 μs |   0.4578 |       - |       - |         - |
|      StringTests | GetPropertyValue |           ? |       1000 |          ? |            ? |    31.920 μs |  0.6019 μs |  1.1306 μs |    32.167 μs |   5.4321 |       - |       - |         - |
|      **StringTests** |       **LookupByPK** |           **?** |         **20** |          **?** |            **?** |    **23.049 μs** |  **0.1991 μs** |  **0.1765 μs** |    **23.078 μs** |   **3.1128** |  **0.0610** |  **0.0610** |         **-** |
|      StringTests |       LookupByPK |           ? |        100 |          ? |            ? |    24.954 μs |  0.4255 μs |  0.5824 μs |    24.864 μs |   3.1128 |  0.0610 |  0.0610 |         - |
|      StringTests |       LookupByPK |           ? |       1000 |          ? |            ? |    49.495 μs |  0.4688 μs |  0.4156 μs |    49.586 μs |   3.1128 |  0.0610 |  0.0610 |         - |
|      **ObjectTests** | **SetPropertyValue** |           **?** |          **?** |          **?** |            **?** |     **1.964 μs** |  **0.0390 μs** |  **0.0941 μs** |     **1.992 μs** |        **-** |       **-** |       **-** |         **-** |
|      StringTests | SetPropertyValue |           ? |         20 |          ? |            ? |     4.765 μs |  0.0311 μs |  0.0259 μs |     4.770 μs |        - |       - |       - |         - |
|      StringTests | SetPropertyValue |           ? |        100 |          ? |            ? |     5.926 μs |  0.0342 μs |  0.0286 μs |     5.923 μs |        - |       - |       - |         - |
|      StringTests | SetPropertyValue |           ? |       1000 |          ? |            ? |    22.042 μs |  0.3289 μs |  0.2916 μs |    22.177 μs |        - |       - |       - |         - |
| **BinaryWriteTests** |            **Write** |          **10** |          **?** |        **128** |            **?** |   **267.964 μs** |  **5.3508 μs** |  **7.8431 μs** |   **265.768 μs** |   **0.9766** |       **-** |       **-** |         **-** |
| BinaryWriteTests |            Write |         100 |          ? |        128 |            ? |   318.319 μs |  6.3370 μs |  5.6176 μs |   318.527 μs |   0.9766 |       - |       - |         - |
| BinaryWriteTests |            Write |          10 |          ? |       1024 |            ? |   278.643 μs |  5.4161 μs |  7.2303 μs |   278.492 μs |   0.9766 |       - |       - |         - |
| BinaryWriteTests |            Write |         100 |          ? |       1024 |            ? |   296.376 μs |  1.8211 μs |  1.5207 μs |   296.520 μs |   0.9766 |       - |       - |         - |
| BinaryWriteTests |            Write |          10 |          ? |    1048576 |            ? |   392.837 μs |  7.7963 μs | 14.2560 μs |   385.311 μs |   0.9766 |       - |       - |         - |
| BinaryWriteTests |            Write |         100 |          ? |    1048576 |            ? |   889.901 μs | 28.5218 μs | 83.6495 μs |   899.000 μs |        - |       - |       - |         - |
|  EmptyWriteTests |            Write |           ? |          ? |          ? |            ? |   235.689 μs |  4.6268 μs |  7.7304 μs |   236.280 μs |   0.4883 |       - |       - |         - |
| StringWriteTests |            Write |          10 |          ? |          ? |           20 |   253.478 μs |  4.6987 μs |  4.3951 μs |   254.862 μs |   0.9766 |       - |       - |         - |
| StringWriteTests |            Write |         100 |          ? |          ? |           20 |   276.430 μs |  5.2774 μs |  4.9365 μs |   276.123 μs |   0.9766 |       - |       - |         - |
| StringWriteTests |            Write |          10 |          ? |          ? |          200 |   283.940 μs |  4.7679 μs |  4.8963 μs |   283.598 μs |   0.9766 |       - |       - |         - |
| StringWriteTests |            Write |         100 |          ? |          ? |          200 |   307.889 μs |  2.4780 μs |  2.1967 μs |   307.981 μs |   0.9766 |       - |       - |         - |

**Comment by nirinchev** [comment]
Super low priority, so definitely as a last-ish step, but can you try running those in release mode on a physical device? It got me a little worried that we're taking 2-3 µs to read a property value until I realized it's a debug build. Not expecting an order of magnitude reduction, but if it doesn't go into the sub-µs range, it may be a little concerning.
